File: server/src/collectors/famousPlaces.js
/**
 * Famous Places / OSM POIs — unified collector (infrastructure category)
 *
 * Nationwide sweep of every OSM feature that qualifies as a "famous place" or
 * noteworthy POI in Japan. Covers the full scope of the former Famous Places
 * collector AND the OSM-tourism fallback that used to live inside the Google
 * My Maps collector.
 *
 * Source: OSM Overpass API, tiled across 12 bboxes that cover the whole of
 * Japan (main islands + Hokkaido + Okinawa + Ogasawara). Nodes AND ways; ways
 * return their `center` as the geometry point.
 *
 * Categories included (no wikidata filter — every OSM element with the tag
 * is returned, regardless of whether it has a Wikidata entry):
 *   tourism   = attraction / viewpoint / museum / artwork / theme_park /
 *               zoo / aquarium / gallery
 *   historic  = any historic=* (castle, monument, memorial, ruins,
 *               archaeological_site, shrine, tomb, wayside_shrine, …)
 *   amenity   = place_of_worship / theatre / arts_centre
 *   leisure   = park / garden / nature_reserve
 *   natural   = peak / volcano / waterfall
 *
 * Each feature carries the full set of useful OSM tags so the UI can show
 * complete info (EN/JA/local names, wikidata, wikipedia, commons, website,
 * phone, opening_hours, admission, operator, heritage/UNESCO, elevation,
 * address, description, image, etc.).
 */

import { fetchOverpassTiled } from './_liveHelpers.js';

export function osmPoiOverpassBody(bbox) {
  const lines = [
    // tourism
    `node["tourism"="attraction"](${bbox});`,
    `way["tourism"="attraction"](${bbox});`,
    `node["tourism"="museum"](${bbox});`,
    `way["tourism"="museum"](${bbox});`,
    `node["tourism"="viewpoint"](${bbox});`,
    `node["tourism"="artwork"](${bbox});`,
    `node["tourism"="theme_park"](${bbox});`,
    `way["tourism"="theme_park"](${bbox});`,
    `node["tourism"="zoo"](${bbox});`,
    `way["tourism"="zoo"](${bbox});`,
    `node["tourism"="aquarium"](${bbox});`,
    `way["tourism"="aquarium"](${bbox});`,
    `node["tourism"="gallery"](${bbox});`,

    // historic — any historic=*
    `node["historic"](${bbox});`,
    `way["historic"](${bbox});`,

    // amenity
    `node["amenity"="place_of_worship"](${bbox});`,
    `way["amenity"="place_of_worship"](${bbox});`,
    `node["amenity"="theatre"](${bbox});`,
    `node["amenity"="arts_centre"](${bbox});`,

    // leisure
    `node["leisure"="park"](${bbox});`,
    `way["leisure"="park"](${bbox});`,
    `node["leisure"="garden"](${bbox});`,
    `way["leisure"="garden"](${bbox});`,
    `node["leisure"="nature_reserve"](${bbox});`,
    `way["leisure"="nature_reserve"](${bbox});`,

    // natural
    `node["natural"="peak"](${bbox});`,
    `node["natural"="volcano"](${bbox});`,
    `node["natural"="waterfall"](${bbox});`,
  ];
  return lines.join('');
}

export function osmPoiCategoryOf(tags) {
  if (tags?.tourism) return tags.tourism;
  if (tags?.historic) return tags.historic;
  if (tags?.amenity === 'place_of_worship') {
    return tags.religion ? `place_of_worship:${tags.religion}` : 'place_of_worship';
  }
  if (tags?.amenity) return tags.amenity;
  if (tags?.leisure) return tags.leisure;
  if (tags?.natural) return tags.natural;
  return 'place';
}

function wikipediaUrl(tag) {
  if (!tag || typeof tag !== 'string') return null;
  const idx = tag.indexOf(':');
  if (idx === -1) return `https://en.wikipedia.org/wiki/${encodeURIComponent(tag.replace(/ /g, '_'))}`;
  const lang = tag.slice(0, idx);
  const title = tag.slice(idx + 1).replace(/ /g, '_');
  return `https://${lang}.wikipedia.org/wiki/${encodeURIComponent(title)}`;
}

function commonsUrl(tag) {
  if (!tag || typeof tag !== 'string') return null;
  const t = tag.startsWith('Category:') || tag.startsWith('File:') ? tag : `File:${tag}`;
  return `https://commons.wikimedia.org/wiki/${encodeURIComponent(t.replace(/ /g, '_'))}`;
}

/**
 * Map an Overpass element to a rich GeoJSON POI feature. Exported so other
 * collectors (e.g. googleMyMaps OSM fallback) can reuse the exact same shape.
 */
export function osmPoiMapFeature(el, i, coords) {
  const t = el.tags || {};
  return {
    type: 'Feature',
    geometry: { type: 'Point', coordinates: coords },
    properties: {
      id: `OSM_${el.type}_${el.id}`,
      platform: 'osm_popular_place',
      osm_type: el.type,
      osm_id: el.id,

      // Names
      name: t['name:en'] || t.name || `Place ${i + 1}`,
      name_ja: t.name || null,
      name_en: t['name:en'] || null,
      name_local: t.name || null,
      alt_name: t.alt_name || null,
      official_name: t.official_name || null,

      // Categorisation
      category: osmPoiCategoryOf(t),
      tourism: t.tourism || null,
      historic: t.historic || null,
      amenity: t.amenity || null,
      leisure: t.leisure || null,
      natural: t.natural || null,
      religion: t.religion || null,
      denomination: t.denomination || null,

      // Identifiers / links
      wikidata: t.wikidata || null,
      wikidata_url: t.wikidata ? `https://www.wikidata.org/wiki/${t.wikidata}` : null,
      wikipedia: t.wikipedia || null,
      wikipedia_url: wikipediaUrl(t.wikipedia),
      wikimedia_commons: t.wikimedia_commons || null,
      wikimedia_commons_url: commonsUrl(t.wikimedia_commons),
      website: t.website || t['contact:website'] || null,
      url: t.url || null,
      image: t.image || null,

      // Contact
      phone: t.phone || t['contact:phone'] || null,
      email: t.email || t['contact:email'] || null,

      // Practical info
      opening_hours: t.opening_hours || null,
      fee: t.fee || null,
      charge: t.charge || null,
      admission: t.admission || null,
      operator: t.operator || null,
      start_date: t.start_date || null,
      heritage: t.heritage || null,
      heritage_operator: t['heritage:operator'] || null,
      unesco: t.unesco || null,

      // Geography
      ele: t.ele || null,
      addr_full: t['addr:full'] || null,
      addr_city: t['addr:city'] || null,
      addr_prefecture: t['addr:province'] || t['addr:state'] || null,
      addr_postcode: t['addr:postcode'] || null,

      // Freeform
      description: t.description || t['description:en'] || null,

      source: 'osm_overpass',
    },
  };
}

export default async function collectFamousPlaces() {
  const features = await fetchOverpassTiled(
    osmPoiOverpassBody,
    osmPoiMapFeature,
    { queryTimeout: 180, timeoutMs: 90_000 },
  );

  const list = features || [];
  return {
    type: 'FeatureCollection',
    features: list,
    _meta: {
      source: 'famous-places',
      fetchedAt: new Date().toISOString(),
      recordCount: list.length,
      live: list.length > 0,
      live_source: list.length > 0 ? 'osm_overpass' : null,
      description: 'Famous places & noteworthy OSM POIs across Japan — tourism/historic/worship/leisure/natural (nationwide tiled Overpass, full metadata)',
    },
    metadata: {},
  };
}

File: native/collectors/sources/famous_places.c
/* collectors/tourism/sources/famous_places.c
 * Port of server/src/collectors/famousPlaces.js (fetchOverpassTiled). Single
 * nationwide tiled Overpass sweep of tourism/historic/worship/leisure/natural
 * POIs with full OSM metadata. osmPoiOverpassBody / osmPoiMapFeature /
 * osmPoiCategoryOf / wikipediaUrl / commonsUrl reproduced verbatim. Properties
 * built in EXACT JS key order (featureUid parity). */
#include "lib/geojson.h"
#include "source.h"
#include "lib/overpass.h"
#include <stdio.h>
#include <string.h>

static void body(const char *b, char *o, size_t n, void *ud) {
  snprintf(o, n,
    "node[\"tourism\"=\"attraction\"](%s);way[\"tourism\"=\"attraction\"](%s);"
    "node[\"tourism\"=\"museum\"](%s);way[\"tourism\"=\"museum\"](%s);"
    "node[\"tourism\"=\"viewpoint\"](%s);"
    "node[\"tourism\"=\"artwork\"](%s);"
    "node[\"tourism\"=\"theme_park\"](%s);way[\"tourism\"=\"theme_park\"](%s);"
    "node[\"tourism\"=\"zoo\"](%s);way[\"tourism\"=\"zoo\"](%s);"
    "node[\"tourism\"=\"aquarium\"](%s);way[\"tourism\"=\"aquarium\"](%s);"
    "node[\"tourism\"=\"gallery\"](%s);"
    "node[\"historic\"](%s);way[\"historic\"](%s);"
    "node[\"amenity\"=\"place_of_worship\"](%s);way[\"amenity\"=\"place_of_worship\"](%s);"
    "node[\"amenity\"=\"theatre\"](%s);"
    "node[\"amenity\"=\"arts_centre\"](%s);"
    "node[\"leisure\"=\"park\"](%s);way[\"leisure\"=\"park\"](%s);"
    "node[\"leisure\"=\"garden\"](%s);way[\"leisure\"=\"garden\"](%s);"
    "node[\"leisure\"=\"nature_reserve\"](%s);way[\"leisure\"=\"nature_reserve\"](%s);"
    "node[\"natural\"=\"peak\"](%s);"
    "node[\"natural\"=\"volcano\"](%s);"
    "node[\"natural\"=\"waterfall\"](%s);",
    b, b, b, b, b, b, b, b, b, b, b, b, b, b, b, b, b, b, b,
    b, b, b, b, b, b, b, b, b);
}

/* osmPoiCategoryOf — writes the category string into buf (cap n). */
static const char *cat_of(cJSON *el, char *buf, size_t n) {
  const char *tour = ov_tag(el, "tourism");
  if (tour) return tour;
  const char *hist = ov_tag(el, "historic");
  if (hist) return hist;
  const char *am = ov_tag(el, "amenity");
  if (am && strcmp(am, "place_of_worship") == 0) {
    const char *rel = ov_tag(el, "religion");
    if (rel) { snprintf(buf, n, "place_of_worship:%s", rel); return buf; }
    return "place_of_worship";
  }
  if (am) return am;
  const char *leis = ov_tag(el, "leisure");
  if (leis) return leis;
  const char *nat = ov_tag(el, "natural");
  if (nat) return nat;
  return "place";
}

/* JS encodeURIComponent only escapes a small set; OSM wiki/commons tags rarely
 * contain those chars. We URL-encode spaces (→ JS replace(/ /g,'_') then
 * encodeURIComponent on a token w/o spaces is identity for typical titles). */
static void wikipedia_url(const char *tag, char *buf, size_t n) {
  if (!tag) { buf[0] = 0; return; }
  const char *colon = strchr(tag, ':');
  if (!colon) {
    snprintf(buf, n, "https://en.wikipedia.org/wiki/");
    size_t l = strlen(buf);
    for (const char *p = tag; *p && l + 1 < n; p++)
      buf[l++] = (*p == ' ') ? '_' : *p;
    buf[l] = 0;
    return;
  }
  int lang_len = (int)(colon - tag);
  snprintf(buf, n, "https://%.*s.wikipedia.org/wiki/", lang_len, tag);
  size_t l = strlen(buf);
  for (const char *p = colon + 1; *p && l + 1 < n; p++)
    buf[l++] = (*p == ' ') ? '_' : *p;
  buf[l] = 0;
}

static void commons_url(const char *tag, char *buf, size_t n) {
  if (!tag) { buf[0] = 0; return; }
  char tmp[512];
  if (strncmp(tag, "Category:", 9) == 0 || strncmp(tag, "File:", 5) == 0)
    snprintf(tmp, sizeof tmp, "%s", tag);
  else
    snprintf(tmp, sizeof tmp, "File:%s", tag);
  snprintf(buf, n, "https://commons.wikimedia.org/wiki/");
  size_t l = strlen(buf);
  for (const char *p = tmp; *p && l + 1 < n; p++)
    buf[l++] = (*p == ' ') ? '_' : *p;
  buf[l] = 0;
}

/* Add `key`: t[a] || t[b] || null  (b may be NULL for single-tag). */
static void add_or_null(cJSON *p, const char *key, cJSON *el,
                        const char *a, const char *b) {
  const char *v = ov_tag(el, a);
  if (!v && b) v = ov_tag(el, b);
  cJSON_AddItemToObject(p, key,
                        v ? cJSON_CreateString(v) : cJSON_CreateNull());
}

static cJSON *map(cJSON *el, int i, double lon, double lat, void *ud) {
  cJSON *f = gj_point_feature(lon, lat);

  cJSON *p = cJSON_CreateObject();
  cJSON *etype = cJSON_GetObjectItem(el, "type");
  cJSON *eid = cJSON_GetObjectItem(el, "id");
  const char *ets = (etype && cJSON_IsString(etype)) ? etype->valuestring : "";
  long long eidv = (eid && cJSON_IsNumber(eid)) ? (long long)eid->valuedouble : 0;
  char idbuf[96];
  snprintf(idbuf, sizeof idbuf, "OSM_%s_%lld", ets, eidv);
  cJSON_AddStringToObject(p, "id", idbuf);
  cJSON_AddStringToObject(p, "platform", "osm_popular_place");
  cJSON_AddStringToObject(p, "osm_type", ets);
  cJSON_AddItemToObject(p, "osm_id", cJSON_CreateNumber((double)eidv));

  /* Names */
  const char *nen = ov_tag(el, "name:en");
  const char *nm = ov_tag(el, "name");
  if (nen)      cJSON_AddStringToObject(p, "name", nen);
  else if (nm)  cJSON_AddStringToObject(p, "name", nm);
  else {
    char nb[32]; snprintf(nb, sizeof nb, "Place %d", i + 1);
    cJSON_AddStringToObject(p, "name", nb);
  }
  cJSON_AddItemToObject(p, "name_ja",
                        nm ? cJSON_CreateString(nm) : cJSON_CreateNull());
  cJSON_AddItemToObject(p, "name_en",
                        nen ? cJSON_CreateString(nen) : cJSON_CreateNull());
  cJSON_AddItemToObject(p, "name_local",
                        nm ? cJSON_CreateString(nm) : cJSON_CreateNull());
  add_or_null(p, "alt_name", el, "alt_name", NULL);
  add_or_null(p, "official_name", el, "official_name", NULL);

  /* Categorisation */
  char catbuf[256];
  cJSON_AddStringToObject(p, "category", cat_of(el, catbuf, sizeof catbuf));
  add_or_null(p, "tourism", el, "tourism", NULL);
  add_or_null(p, "historic", el, "historic", NULL);
  add_or_null(p, "amenity", el, "amenity", NULL);
  add_or_null(p, "leisure", el, "leisure", NULL);
  add_or_null(p, "natural", el, "natural", NULL);
  add_or_null(p, "religion", el, "religion", NULL);
  add_or_null(p, "denomination", el, "denomination", NULL);

  /* Identifiers / links */
  const char *wd = ov_tag(el, "wikidata");
  cJSON_AddItemToObject(p, "wikidata",
                        wd ? cJSON_CreateString(wd) : cJSON_CreateNull());
  if (wd) {
    char wb[160]; snprintf(wb, sizeof wb, "https://www.wikidata.org/wiki/%s", wd);
    cJSON_AddStringToObject(p, "wikidata_url", wb);
  } else cJSON_AddItemToObject(p, "wikidata_url", cJSON_CreateNull());
  const char *wp = ov_tag(el, "wikipedia");
  cJSON_AddItemToObject(p, "wikipedia",
                        wp ? cJSON_CreateString(wp) : cJSON_CreateNull());
  char wpu[600]; wikipedia_url(wp, wpu, sizeof wpu);
  cJSON_AddItemToObject(p, "wikipedia_url",
                        wpu[0] ? cJSON_CreateString(wpu) : cJSON_CreateNull());
  const char *wc = ov_tag(el, "wikimedia_commons");
  cJSON_AddItemToObject(p, "wikimedia_commons",
                        wc ? cJSON_CreateString(wc) : cJSON_CreateNull());
  char wcu[600]; commons_url(wc, wcu, sizeof wcu);
  cJSON_AddItemToObject(p, "wikimedia_commons_url",
                        wcu[0] ? cJSON_CreateString(wcu) : cJSON_CreateNull());
  add_or_null(p, "website", el, "website", "contact:website");
  add_or_null(p, "url", el, "url", NULL);
  add_or_null(p, "image", el, "image", NULL);

  /* Contact */
  add_or_null(p, "phone", el, "phone", "contact:phone");
  add_or_null(p, "email", el, "email", "contact:email");

  /* Practical info */
  add_or_null(p, "opening_hours", el, "opening_hours", NULL);
  add_or_null(p, "fee", el, "fee", NULL);
  add_or_null(p, "charge", el, "charge", NULL);
  add_or_null(p, "admission", el, "admission", NULL);
  add_or_null(p, "operator", el, "operator", NULL);
  add_or_null(p, "start_date", el, "start_date", NULL);
  add_or_null(p, "heritage", el, "heritage", NULL);
  add_or_null(p, "heritage_operator", el, "heritage:operator", NULL);
  add_or_null(p, "unesco", el, "unesco", NULL);

  /* Geography */
  add_or_null(p, "ele", el, "ele", NULL);
  add_or_null(p, "addr_full", el, "addr:full", NULL);
  add_or_null(p, "addr_city", el, "addr:city", NULL);
  add_or_null(p, "addr_prefecture", el, "addr:province", "addr:state");
  add_or_null(p, "addr_postcode", el, "addr:postcode", NULL);

  /* Freeform */
  add_or_null(p, "description", el, "description", "description:en");

  cJSON_AddStringToObject(p, "source", "osm_overpass");
  cJSON_AddItemToObject(f, "properties", p);
  return f;
}

static int run(const source_ctx *ctx, intel_sink *sink) {
  int n = overpass_tiled_collect(ctx, sink, body, 180, 90000, map, NULL);
  return n >= 0 ? 0 : -1;
}

static const source_def famous_places_def = {
  .id = "famous-places", .collector = "tourism", .name = "OSM Famous Places",
  .name_ja = "OSM 名所", 
  .update_interval_sec = 604800, .run = run };
REGISTER_SOURCE(famous_places_def)
